lib/delta-types/src/Data/Delta/Embedding.hs

```diff
@@ -0,0 +1,256 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+
+{-|
+Copyright: © 2021-2023 IOHK, 2024 Cardano Foundation
+License: Apache-2.0
+
+Embeddings of delta types.
+-}
+module Data.Delta.Embedding (
+    -- $Embedding
+    -- * Embedding
+      Embedding
+    , module Data.Semigroupoid
+    , Embedding' (..)
+    , mkEmbedding
+    , fromEmbedding
+    , pair
+    , liftUpdates
+    , replaceFromApply
+
+    -- * Internal
+    , inject
+    , project
+    , Machine (..)
+    , idle
+    , pairMachine
+    , fromState
+    ) where
+
+import Prelude
+
+import Control.Exception
+    ( SomeException
+    )
+import Data.Delta.Core
+    ( Delta (..)
+    , Replace (..)
+    )
+import Data.Either
+    ( fromRight
+    )
+import Data.Semigroupoid
+    ( Semigroupoid (..)
+    )
+
+{-------------------------------------------------------------------------------
+    Embedding
+-------------------------------------------------------------------------------}
+{- $Embedding
+
+An 'Embedding'@ da db@ embeds one type and its delta type @da@
+into another type and its delta type @db@.
+
+For reasons of efficiency, 'Embedding' is an abstract type.
+It is constructed using the 'Embedding'' type, which has
+three components.
+
+* 'write' embeds values from the type @a = 'Base' da@
+    into the type @b = 'Base' db@.
+* 'load' attempts to retrieve the value of type @a@
+    from the type @b@, but does not necessarily succeed.
+* 'update' maps a delta type @da@ to a delta type @db@.
+    For this mapping, the value of type @a@ and a corresponding
+    value of type @b@ are provided;
+    the delta types @da@ and @db@ are
+    relative to these values.
+
+The embedding of one type into the other is characterized by the following
+properties:
+
+* The embedding need __not__ be __surjective__:
```

Comment:
```suggestion
* The embedding needs to be  __not__ surjective__:
```

Reply:
doesn't need to be ?

Reply:
It can be surjective, but does not need to be. I have changed the phrasing to

> not necessarily surjective